src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,125 @@
+  ) => Promise<SwapRoute>
+}
+export class TenderlyProvider implements ISimulator {
+  private tenderlyBaseUrl: string
+  private tenderlyUser: string
+  private tenderlyProject: string
+  private tenderlyAccessKey: string
+
+  constructor(tenderlyBaseUrl: string, tenderlyUser: string, tenderlyProject: string, tenderlyAccessKey: string) {
+    this.tenderlyBaseUrl = tenderlyBaseUrl
+    this.tenderlyUser = tenderlyUser
+    this.tenderlyProject = tenderlyProject
+    this.tenderlyAccessKey = tenderlyAccessKey
+  }
+
+  public async simulateTransaction(
+    tokenIn: Token,
+    fromAddress: string,
+    route: SwapRoute,
+    poolProvider: IV3PoolProvider,
+  ): Promise<SwapRoute> {
+
+    const { calldata } = route.methodParameters!
+    if(!route.methodParameters) {
+      throw new Error("No calldata provided to simulate transaction")
+    }
+    log.info(
+      {
+        calldata: route.methodParameters.calldata,
+        fromAddress: fromAddress,
+        chainId: tokenIn.chainId,
+        tokenInAddress: tokenIn.address,
+      },
+      'Simulating transaction via Tenderly'
+    )
+
+    const approve = {
+      network_id: tokenIn.chainId,
+      input: APPROVE_TOKEN_FOR_TRANSFER,
+      to: tokenIn.address,
+      value: "0",
+      from: fromAddress,
+      gasPrice: "0",
+      gas: 30000000,
+    }
+
+    const swap = {
+      network_id: tokenIn.chainId,
+      input: calldata,
+      to: V3_ROUTER2_ADDRESS,
+      value: "0",
+      from: fromAddress,
+      gasPrice: "0",
+      gas: 30000000,
+      type: 1,
+      estimate_gas: true,
+    }
+
+    const body = {"simulations": [approve, swap]}
+    const opts = {
+      headers: {
+        'X-Access-Key': this.tenderlyAccessKey,
+      },
+    }
+    const url = TENDERLY_BATCH_SIMULATE_API(this.tenderlyBaseUrl, this.tenderlyUser, this.tenderlyProject)
+    const resp = await axios.post(url, body, opts)
+
+    // Validate tenderly response body
+    if(resp.data && resp.data.simulation_results.length == 2 && resp.data.simulation_results[1].transaction && !resp.data.simulation_results[1].transaction.error_message) {
+      log.info({approve:resp.data.simulation_results[0],swap:resp.data.simulation_results[1]}, 'Simulated Transaction Via Tenderly')
+
+      // Parse the gas used in the simulation response object, and then pad it so that we overestimate.
+      const gasUsed = resp.data.simulation_results[1].transaction.gas_used * ESTIMATE_MULTIPLIER
+      const l1FeeInWei = BigNumber.from((gasUsed).toFixed())
+      const { gasCostL1QuoteToken, gasCostL1USD } = await getGasCostsInUSDandQuote(tokenIn, l1FeeInWei, poolProvider)
+      // gasUsedL1 is the gas units used calculated from the bytes of the calldata
+      // gasCostL1USD and gasCostL1QuoteToken is the cost of gas in each of those tokens
+      route = {...route, estimatedGasUsed:l1FeeInWei, estimatedGasUsedUSD: gasCostL1USD, estimatedGasUsedQuoteToken: gasCostL1QuoteToken}
+      return route
```

Comment:
What about quoteGasAdjusted

Also this doesn't seem to support exactIn/exactOut correctly

- for exactIn the quote token is tokenOut (i.e. user specifies an amount of token in, auto routers returns their quote for receiving token out)
- for exactOut the quote token is tokenIn (i.e. user specifies an amount of token out, auto router returns their quote for receiving token in)